risc0/zkvm/src/receipt.rs

```diff
@@ -149,24 +150,64 @@ pub struct ReceiptMetadata {
     pub output: Digest,
 }
 
+/// The [SessionReceipt] is implemented for receipts that represent
+/// [crate::Sessions]
+///
+/// There are several different flavors of receipts and this trait defines the
+/// the functions that all [Session] receipts have such as accessing the
+/// journal and a way to verify the receipt.
+pub trait SessionReceipt: Debug {
+    /// Verifies the integrity of this receipt.
+    ///
+    /// Uses the ZKP system to cryptographically verify that each constituent
+    /// Segment has a valid receipt, and validates that these [SegmentReceipt]s
+    /// stitch together correctly, and that the initial memory image matches the
+    /// given `image_id` parameter.
+    #[cfg(not(target_os = "zkvm"))]
+    #[must_use]
+    fn verify(&self, image_id: Digest) -> Result<(), VerificationError>;
+
+    /// All session receipts have a journal
+    fn get_journal(&self) -> &Vec<u8>;
+
+    /// this provides a way to serialize a receipt
+    fn encode(&self) -> Vec<u8>;
+
+    /// this is used for downcasting, primarily used for testing
+    fn as_any(&self) -> &dyn Any;
```

Comment:
I'd really like to drop this somehow. can we add more methods to the trait so we don't need to downcast?

Reply:
How about we do `#[cfg(test)]` so that it's used for testing purposes only? I think it's useful to be able to downcast for tests so that we can more freely dive into testing the `FlatSessionReceipt`

Reply:
And we could remove it all together once we implement the combined receipt